src/hotspot/share/compiler/compilationMemStatInternals.hpp

```diff
@@ -0,0 +1,285 @@
+private:
+  template <typename T, typename dT>
+  struct C {
+    T start, peak, cur;
+    void init(T v)        { start = cur = peak = v; }
+    void update(T v)      { cur = v; if (v > peak) peak = v; }
+    dT end_delta() const  { return (dT)cur - (dT)start; }
+    size_t temporary_peak_size() const { return MIN2(peak - cur, peak - start); }
+  };
+  struct Entry {
+    PhaseInfo info;
+    int level;
+    C<size_t, ssize_t> _bytes;
+    C<unsigned, signed int> _live_nodes;
+  };
+  SimpleFifo<Entry, max_num_phases> _fifo;
+  DEBUG_ONLY(bool _inbetween_phases;)
+public:
+  FootprintTimeline();
+  void copy_from(const FootprintTimeline& other);
+  inline void on_footprint_change(size_t cur_abs, unsigned cur_nodes);
+  void on_phase_end(size_t cur_abs, unsigned cur_nodes);
+  void on_phase_start(PhaseInfo info, size_t cur_abs, unsigned cur_nodes, int level);
+  void print_on(outputStream* st) const;
+};
+
+// We keep the name of the involved symbols in Symbol (made permanent) instead of resolving them to string and
+// storing those. That significantly reduces footprint for the result store and delays resolving until printing
+// time, which may be never.
+class FullMethodName {
+  Symbol* _k;
+  Symbol* _m;
+  Symbol* _s;
+public:
+  FullMethodName();
+  FullMethodName(const Method* m);
+  FullMethodName(const FullMethodName& o);
+  FullMethodName& operator=(const FullMethodName& o);
+  void make_permanent();
+  void print_on(outputStream* st) const;
+  char* as_C_string(char* buf, size_t len) const;
+  bool operator== (const FullMethodName& b) const;
+  DEBUG_ONLY(bool is_test_class() const;)
+};
+
+// ArenaState is the central data structure holding all statistics and temp data during
+// a single compilation. It is created on demand (if memstat is active) and tied to the
+// CompilerThread.
+class ArenaStatCounter : public CHeapObj<mtCompiler> {
+
+  FullMethodName _fmn;
+
+  // from directives
+  const bool _should_print_memstat;
+  const bool _should_crash_on_memlimit;
+
+  // Bytes total now
+  size_t _current;
+  // Bytes total at last global peak
+  size_t _peak;
+  // Bytes per arena/phase, now
+  ArenaCounterTable _counters_current;
+  // Bytes per arena/phase when we last reached the global peak
+  ArenaCounterTable _counters_at_global_peak;
+
+  // Number of live nodes now (C2 only)
+  unsigned _live_nodes_current;
+  // Number of live nodes at global peak (C2 only)
+  unsigned _live_nodes_at_global_peak;
+
+  // MemLimit handling
+  const size_t _limit;
+  bool _hit_limit;
+  bool _limit_in_process;
+
+  // Keep track of current C2 phase
+  int _phase_counter;
+  PhaseInfoStack _phase_info_stack;
+
+  // Keep track of C2 phase allocations over time
+  FootprintTimeline _timeline;
+
+  const CompilerType _comp_type;
+  const int _comp_id;
+
+  DEBUG_ONLY(bool _is_test_class;)
+
+  int retrieve_live_node_count() const;
+
+  DEBUG_ONLY(void verify() const;)
```

Comment:
Unused. I suggest to either call this function from some appropriate point (in debug mode only) or just remove it.

Reply:
I'll use it.